From 0d8394ee5c52e97e82ebe4641cf0d9ebcbe147ff Mon Sep 17 00:00:00 2001
From: Rui Miguel Silva <rui.silva@linaro.org>
Date: Tue, 6 Sep 2022 16:47:06 +0100
Subject: [PATCH] Add atomic outline to fix build for opemamp

Add memory model 5 atomic ouline support (_sync) to fix
missing symbol when compiling with recent gcc (12.2).

Signed-off-by: Rui Miguel Silva <rui.silva@linaro.org>

Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Rui Miguel Silva <rui.silva@arm.com>
---
 deployments/se-proxy/opteesp/lse.S | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/deployments/se-proxy/opteesp/lse.S b/deployments/se-proxy/opteesp/lse.S
index 840683a6671a..8e466d65fc2b 100644
--- a/deployments/se-proxy/opteesp/lse.S
+++ b/deployments/se-proxy/opteesp/lse.S
@@ -5,6 +5,7 @@
 
 .text
 .globl __aarch64_cas4_acq_rel
+.globl __aarch64_cas4_sync
 
 __aarch64_cas4_acq_rel:
 	mov	w16, w0
@@ -16,4 +17,12 @@ __aarch64_cas4_acq_rel:
 	cbnz	w17, 0b
 1:	ret
 
+__aarch64_cas4_sync:
+	mov	w16, w0
+	ldxr	w0, [x2]
+	cmp	w0, w16
+0:	bne	1f
 
+	stlxr	w17, w1, [x2]
+	cbnz	w17, 0b
+1:	ret
-- 
2.37.3

